### ISSUE TITLE
Remote signing support

### ISSUE THREAD

**Issue by TommyLike** [rpm-rs/rpm#80]
**Background**
I am trying to do a remote pgp sign by utilizing this rpm package, but it seems the header and signature functions are public to crate only, how can we read&update these sections outside of rpm package?
```
package.metadata.header.**write**(&mut header_bytes)?;
package.metadata.signature = Header::<IndexSignatureTag>::**new_signature_header**();
```

**Comment by TommyLike** [comment]
@drahnr @dralley  could you take a look?

**Comment by dralley** [comment]
There is a `sign()` method on `Package` which is public, is that not usable in your case?

**Comment by TommyLike** [comment]
@dralley yeah, what I need is to split the content and do a remote signature then assemble them back into signature.

**Comment by drahnr** [comment]
So from what I understand, you send the hash to a remote entity which ows your signing key rather than signing it locally?

I'd probably recommend to implement a custom `Signer` https://github.com/rpm-rs/rpm/blob/41723d969242c61b8e752a929e905730a18f2590/src/rpm/signature/traits.rs#L27 where the actual signing is deferred, so https://github.com/rpm-rs/rpm/blob/41723d969242c61b8e752a929e905730a18f2590/src/rpm/signature/pgp.rs#L44 would be `None` and you'd have to calculate the hash locally, send it to your signing entity and return it.

**Comment by dralley** [comment]
@TommyLike I imagine the idea here is to implement obs-sign, or something like it, on your own - yes?  That makes sense.

**Comment by TommyLike** [comment]
@drahnr yeah, I am trying to create rust version of obs sign with more efficient, secure and e2e solution.

**Comment by drahnr** [comment]
But then the method of choice should be to implement the mentioned traits. If you have trouble with that, or the API is insufficient, please open another issue :)


I am not sure there is anything to do given the generality of the issue at hand?

**Comment by TommyLike** [comment]
> But then the method of choice should be to implement the mentioned traits. If you have trouble with that, or the API is insufficient, please open another issue :)
> 
> I am not sure there is anything to do given the generality of the issue at hand?

the question is in order to sign multiple rpms concurrently we seperated the signing process/trait into three different workers: split file->do remote sign->assemble signature in which we can't simply impl them all in this trait.

**Comment by drahnr** [comment]
'worker' meaning individual vms/nodes/machines? And that cannot be changed/re-partitioned?

My point is: you don't need to split, you don't need to assemble. All you need to do is _sign_ the thing. If pkg content is living in a DB, then we'd need more info.

But to be frank, it feels like you're trying to form the cold end of the iron.

**Comment by dralley** [comment]
@drahnr I believe the basic idea of obs-sign (which I am not fully familiar with, so anyone please feel free to correct me) that you have a client which opens the RPM and makes a checksum of the relevant bits.  The client can then pass only this checksum to an external service which manages the signing.  The service then returns a signature which the client can inject into the RPM.

This approach avoids uploading the RPM contents themselves and keeps the signing key management contained to a single designated machine, which is easier to lock down.

Imagine a build system where you have multiple build servers.  Putting the signing keys on each individual build server opens up more attack surface and opportunity for those keys to be compromised.  You could probably dump all the unsigned RPMs onto a shared NFS drive, or upload full RPMs to a signing service, but maybe that involves too much data movement and/or potential for needlessly sharing sensitive data?  So instead you just pass around the checksum.

@TommyLike Is that a good summary?

**Comment by TommyLike** [comment]
> onto a shared NFS drive, or upload full RPMs to a signing service, bu

@dralley exactly 😄 , much appreciated for your explanation.

**Comment by TommyLike** [comment]
@drahnr sure
1. this is the changes I made from official rpm-rs project: https://github.com/rpm-rs/rpm/compare/master...TommyLike:rpm-rs:master
2.  and this is how we utilize the updated rpm-rs in our demo project:
https://github.com/TommyLike/signatrust/blob/main/src/client/file_handler/rpm.rs check the `split_data` and `assemble_data` functions

**Comment by dralley** [comment]
@TommyLike If you do plan to work on this could you take it as assigned?

**Comment by TommyLike** [comment]
@dralley  sure, will create PR when ready.

**Comment by dralley** [comment]
@TommyLike In https://github.com/rpm-rs/rpm/pull/225 I'm removing all of the implementation that uses legacy signatures and digests, which includes the header + payload ones.  That simplifies the problem quite a bit.

I *think* what you should be able to do is pull the sha256 checksum of the header out of the signature header, and then sign it, then build your new signature header with the sha256 checksum and the signature.

But if you like, I can look at making a function available to just serialize the header so that you can generate whatever checksum you want

**Comment by dralley** [comment]
@TommyLike It sounds like, from the discussion on the issue you filed (https://github.com/rpgp/rpgp/issues/225), that there's not a way to send only the digest of the header, practically speaking you do still need to actually send the header.

Modern RPMs don't need header + payload signing though, so the header bytes are all you need.  And as mentioned in the last comment `rpm-rs` no longer creates header + payload signatures anymore.

You can probably simplify https://github.com/TommyLike/signatrust/ keeping this in mind.

Do you have recommendations on how to streamline creating a custom `Signer` that accommodates your use cases and avoids needing to roll anything on your own?  Otherwise I think we can close this out.
